### PR TITLE
2pc review

### PR DESCRIPTION
**Pull request by markhannum**
Logic for 2-phased commit

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>8/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover_noroll1_generated<br>queuedb_rollover<br>lostwrite<br>dupewrite<br>cdb2api_2pc<br>tunables</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>10/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>sc_transactional_rowlocks_generated<br>scindex<br>phys_rep_tiered<br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover<br>queuedb_rollover_noroll1_generated<br>dupewrite<br>cdb2api_2pc</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>162/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>sc_transactional_rowlocks_generated<br>scindex<br>phys_rep_tiered<br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover_noroll1_generated<br>queuedb_rollover<br>sp<br>dupewrite</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>6/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>2pc<br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover<br>dupewrite</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>8/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>phys_rep<br>scindex<br>phys_rep_tiered<br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover_noroll1_generated<br>dupewrite</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>9/511 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_transactional_rowlocks_generated<br>phys_rep<br>scindex<br>phys_rep_tiered<br>sc_resume_logicalsc_generated<br>sc_resume<br>queuedb_rollover_noroll1_generated<br>queuedb_rollover<br>dupewrite</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ bdb/rep.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>5/512 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>phys_rep_tiered<br>queuedb_rollover_noroll1_generated<br>dupewrite<br>tunables</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ bdb/rep.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>2/512 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>cldeadlock<br>dupewrite</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ bdb/rep.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/toblock.c @@
+ Please make sure that a mutex is unlocked before returning.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>5/513 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>2pc<br>scindex<br>queuedb_rollover_noroll1_generated<br>incoherent_slow<br>dupewrite</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ berkdb/txn/txn_util.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ bdb/rep.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlsession.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/disttxn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/osqlblockproc.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/toblock.c @@
+ Please make sure that a mutex is unlocked before returning.
```

File: db/toblock.c [hunk 1]
```diff
@@ -5562,46 +5648,187 @@ static int toblock_main_int(struct javasp_trans_state *javasp_trans_handle,
...
+                                                         iq->sorese->coordinator_tier, iq->sorese->coordinator_master);
+                                }
+
+                                /* We have committed dist-txn durably: commit & replicate */
+                                if (waitrc == HAS_COMMITTED) {
+                                    Pthread_mutex_lock(&blklk);
+                                    prepared_count--;
+                                    Pthread_mutex_unlock(&blklk);
+                                    irc = trans_commit_adaptive(iq, parent_trans, source_host);
+                                    if (iq->sorese->is_coordinator) {
```
Comment (riverszhang89) [inline db/toblock.c:5704]: note to myself- The "wait-for-propagation" logic is here

File: db/toblock.c [hunk 2]
```diff
@@ -5562,46 +5648,187 @@ static int toblock_main_int(struct javasp_trans_state *javasp_trans_handle,
...
+                                        participant_wait(iq->sorese->dist_txnid, iq->sorese->coordinator_dbname,
+                                                         iq->sorese->coordinator_tier, iq->sorese->coordinator_master);
+                                }
+
+                                /* We have committed dist-txn durably: commit & replicate */
+                                if (waitrc == HAS_COMMITTED) {
+                                    Pthread_mutex_lock(&blklk);
+                                    prepared_count--;
+                                    Pthread_mutex_unlock(&blklk);
+                                    irc = trans_commit_adaptive(iq, parent_trans, source_host);
```
Comment (riverszhang89) [inline db/toblock.c:5703]: note to myself - commit here

File: db/disttxn.c [hunk 3]
```diff
@@ -0,0 +1,2268 @@
...
+            snprintf(errmsg, errmsglen, "failed writing to disttxn table");
+            start_timer = comdb2_time_epochms();
+            abort_transaction_lk(dtran, 0);
+            stop_timer = comdb2_time_epochms();
+            if (gbl_debug_disttxn_trace) {
+                logmsg(LOGMSG_USER, "DISTTXN %s abort txn took %d ms\n", __func__, stop_timer - start_timer);
+            }
+        } else {
+            start_timer = comdb2_time_epochms();
+            commit_participants_lk(dtran);
```
Comment (riverszhang89) [inline db/disttxn.c:1330]: note to myself- now tell participants to commit

File: db/sltdbt.c [hunk 4]
```diff
@@ -162,6 +163,10 @@ static int handle_op_local(struct ireq *iq, int (*init)(struct ireq *),
             goto done;
     }
 
+    if (iq->sorese && iq->sorese->is_coordinator) {
+        dispatch_participants(iq->sorese->dist_txnid);
```
Comment (riverszhang89) [inline db/sltdbt.c:167]: note to myself - prepare here

**Review by riverszhang89 — APPROVED** [review]

**Review by riverszhang89 — APPROVED** [review]
Looks good to merge!

Do records in comdb2_distributed_transactions get deleted? Maybe we can extend the table with an "lsn-file" field, that records the current logfile, and then log-delete code can also delete records in this table if they're past MIN_KEEP_LOGS.

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>3/513 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>cldeadlock<br>sc_truncate_lockorder_generated<br>dupewrite</p>

**Review by riverszhang89 — APPROVED** [review]

**Pull request merged**